packages/blocks/src/root-block/widgets/scroll-anchoring/index.ts

```diff
@@ -0,0 +1,234 @@
+import type { DocMode } from '@blocksuite/affine-model';
+
+import { requestConnectedFrame } from '@blocksuite/affine-shared/utils';
+import { WidgetComponent } from '@blocksuite/block-std';
+import { Bound, deserializeXYWH } from '@blocksuite/global/utils';
+import { computed, signal } from '@preact/signals-core';
+import { cssVarV2 } from '@toeverything/theme/v2';
+import { css, html, nothing, unsafeCSS } from 'lit';
+import { classMap } from 'lit/directives/class-map.js';
+import { styleMap } from 'lit/directives/style-map.js';
+
+import { EdgelessRootService } from '../../edgeless/edgeless-root-service.js';
+import { ZOOM_MAX } from '../../edgeless/utils/zoom.js';
+
+export const AFFINE_SCROLL_ANCHORING_WIDGET = 'affine-scroll-anchoring-widget';
+
+export class AffineScrollAnchoringtWidget extends WidgetComponent {
+  static override styles = css`
+    :host {
+      pointer-events: none;
+      position: absolute;
+      left: 0px;
+      top: 0px;
+      transform-origin: left top;
+      contain: size layout;
+      z-index: 1;
+
+      & .highlight {
+        position: absolute;
+        box-sizing: border-box;
+
+        &.edgeless {
+          border-width: 1.39px;
+          border-style: solid;
+          border-radius: 8px;
+          border-color: ${unsafeCSS(cssVarV2('primary/color'))};
+          box-shadow: ${unsafeCSS(cssVarV2('active/shadow'))};
+        }
+
+        &.page {
+          border-radius: 5px;
+          background-color: ${unsafeCSS(cssVarV2('hover/color'))};
+        }
+      }
+    }
+  `;
+
+  #requestUpdateFn = () => this.requestUpdate();
+
+  #resizeObserver: ResizeObserver = new ResizeObserver(this.#requestUpdateFn);
+
+  anchor = signal<{ mode: DocMode; id: string } | null>(null);
+
+  anchorBounds = signal<Bound | null>(null);
+
+  highlighted = computed(() => this.service.selectionManager.find('highlight'));
+
+  #getBoundsInEdgeless() {
+    const service = this.std.getOptional(EdgelessRootService);
```

Comment:
Don't get `service` here, you should get [`GfxControllerIdentifier`](https://github.com/toeverything/blocksuite/blob/aedb56ccaeec689b20f2218eccb802584568dd2c/packages/framework/block-std/src/gfx/controller.ts#L196) to get the viewport.